profile.go

```diff
@@ -0,0 +1,199 @@
+	switch t := s.val.(type) {
+	case *url.URL:
+		return t.String(), nil
+	case asn1.ObjectIdentifier:
+		return t.String(), nil
+	default:
+		return "", fmt.Errorf("no valid EAT profile")
+	}
+}
+
+// IsURI checks whether a stored profile is a URI
+func (s Profile) IsURI() bool {
+	switch s.val.(type) {
+	case *url.URL:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsOID checks whether a stored profile is an OID
+func (s Profile) IsOID() bool {
+	switch s.val.(type) {
+	case asn1.ObjectIdentifier:
+		return true
+	default:
+		return false
+	}
+}
+
+// DecodeProfileCBOR decodes from a received CBOR data the profile
+// as either a URL or a Object Identifier
+func (s *Profile) DecodeProfileCBOR(val interface{}) error {
+	switch t := val.(type) {
+	case string:
+		lurl, err := url.Parse(t)
+		if err != nil {
+			return fmt.Errorf("profile URL parsing failed: %w", err)
+		}
+		if !lurl.IsAbs() {
+			return fmt.Errorf("profile URL not in absolute form: %w", err)
+		}
+		s.val = lurl
+	case []byte:
+		var profileOID asn1.ObjectIdentifier
+		_, err := asn1.Unmarshal(t, &profileOID)
+		if err != nil {
+			return fmt.Errorf("malformed profile OID")
+		}
+		s.val = profileOID
+	default:
+		return fmt.Errorf("decoding failed malformed CBOR")
+	}
+	return nil
+}
+
+// MarshalCBOR will encode the Profile value either as a CBOR text string(URL),
+// or as byte array
+func (s Profile) MarshalCBOR() ([]byte, error) {
+	switch t := s.val.(type) {
+	case *url.URL:
+		return em.Marshal(t.String())
+
+	case asn1.ObjectIdentifier:
+		var asn1OID []byte
+		asn1OID, err := asn1.Marshal(t)
+		if err != nil {
+			return nil, fmt.Errorf("asn1 encoding failed for OID: %w", err)
+		}
+		return em.Marshal(asn1OID)
+	default:
+		return nil, fmt.Errorf("invalid type for EAT profile")
+	}
+}
+
+// UnmarshalCBOR attempts to initialize the Profile from the presented
+// CBOR data. The data must be a text string, representing a URL
+// or a byte array representing a Object Identifier
+func (s *Profile) UnmarshalCBOR(data []byte) error {
+	var val interface{}
+	if len(data) == 0 {
+		return fmt.Errorf("decoding of CBOR data failed: zero length data buffer")
+	}
+	if err := dm.Unmarshal(data, &val); err != nil {
+		return fmt.Errorf("decoding of CBOR data failed: %w", err)
+	}
+	if err := s.DecodeProfileCBOR(val); err != nil {
+		return fmt.Errorf("invalid profile data: %w", err)
+	}
+	return nil
+}
+
+func decodeOIDfromString(val string) (asn1.ObjectIdentifier, error) {
+	// Attempt to decode OID from received string
+	var oid asn1.ObjectIdentifier
+	for _, s := range strings.Split(val, ".") {
+		num, err := strconv.ParseInt(s, 10, 32)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract OID from string: %w", err)
+		}
+		n := int(num)
+		oid = append(oid, n)
+	}
+	return oid, nil
+}
+
+// DecodeProfileJSON decodes a valid profile, from the received
+// JSON string, mapping it to either a URL or an OID
+func (s *Profile) DecodeProfileJSON(val string) error {
```

Comment:
this is not a public method
```suggestion
func (s *Profile) decodeProfileJSON(val string) error {
```